Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,53 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+    {
+       return true;
+    }
+    else
+    {
+       return false;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
-}
+    int msRemaining;
+    if (isCharging())
+    {
+      msRemaining = totalAmpHoursUsed() / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+    }
+    else
+    {
+      msRemaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+    }
+
+    return QTime(0, 0, 0, 0).addMSecs(msRemaining);
 
+}
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    if (previousTime_.isNull())
+    {
+        previousTime_ = batteryData.time;
+    }
+    QTime presentTime = batteryData.time;
+    int deltaMseconds = abs(presentTime.msecsTo(previousTime_));
+    double deltaHours = (double)deltaMseconds * MS_TO_HOURS_CONVERSION;
+    previousTime_ = batteryData.time;
+
+    if (previousCurrent_ == NULL)
+    {
+        previousCurrent_ = batteryData.current;
+    }
+    current_ = batteryData.current;
+    averageCurrent_ = (batteryData.current + previousCurrent_) / 2;
+    totalAmpHoursUsed_ += averageCurrent_ * deltaHours;
+    previousCurrent_ = batteryData.current;
```

Comment:
By the end of this function. previousCurrent_ and current_ have the same value and so it's pointless to keep both, what do you think the proper fix for this?